File: Arrays/P022.py
def brute(nums):
    n=len(nums)
    count0=count1=count2=0
    for n in nums:
        if n==0:
            count0+=1
        elif n==1:
            count1+=1
        else:
            count2+=1
    i=0
    for _ in range(count0):
        nums[i]=0
        i+=1
    for _ in range(count1):
        nums[i]=1
        i+=1
    for _ in range(count2):
        nums[i]=2
        i+=1    
    return nums

def optimal(nums):
    n=len(nums)
    low,mid,high=0,0,n-1
    while(mid<=high):
        if nums[mid]==0:
            nums[low],nums[mid]=nums[mid],nums[low]
            low+=1
            mid+=1
        elif nums[mid]==1:
            mid+=1
        else:
            nums[mid],nums[high]=nums[high],nums[mid]
            high-=1
    return nums

if __name__=='__main__':
    nums=input('Enter array: ').split()
    nums=list(map(int,nums))
    print('Brute: ',brute(nums))
    print('Optimal: ',optimal(nums))

File: Striver_C++_Problems/Arrays/P022.cpp
// P022.cpp - Sort array of 0s, 1s, 2s

#include <bits/stdc++.h>
using namespace std;

vector<int> brute(vector<int> nums) {
    int count0 = 0, count1 = 0, count2 = 0;
    for (int n : nums) {
        if (n == 0) count0++;
        else if (n == 1) count1++;
        else count2++;
    }
    int i = 0;
    for (int k = 0; k < count0; k++) nums[i++] = 0;
    for (int k = 0; k < count1; k++) nums[i++] = 1;
    for (int k = 0; k < count2; k++) nums[i++] = 2;
    return nums;
}

vector<int> optimal(vector<int> nums) {
    int n = nums.size();
    int low = 0, mid = 0, high = n - 1;
    while (mid <= high) {
        if (nums[mid] == 0) {
            swap(nums[low], nums[mid]);
            low++;
            mid++;
        } else if (nums[mid] == 1) {
            mid++;
        } else {
            swap(nums[mid], nums[high]);
            high--;
        }
    }
    return nums;
}

int main() {
    cout << "Enter array: ";
    string line;
    getline(cin, line);
    stringstream ss(line);
    vector<int> nums;
    int x;
    while (ss >> x) nums.push_back(x);

    auto b = brute(nums);
    auto o = optimal(nums);

    cout << "Brute:  ";
    for (int v : b) cout << v << " ";
    cout << "\nOptimal: ";
    for (int v : o) cout << v << " ";
    cout << "\n";
    return 0;
}
